File: python/triangle/triangle.py
import typing as tp


def equilateral(sides: tp.Sequence[float]) -> bool:
    return is_triangle(sides) and len(set(sides)) == 1


def isosceles(sides: tp.Sequence[float]) -> bool:
    return is_triangle(sides) and len(set(sides)) <= 2


def scalene(sides: tp.Sequence[float]) -> bool:
    return is_triangle(sides) and len(set(sides)) == 3


def is_triangle(sides: tp.Sequence[float]) -> bool:
    try:
        x, y, max_ = sorted(sides)
    except ValueError:
        return False
    return x + y > max_

File: cpp/triangle/triangle.cpp
#include "triangle.h"
#include <stdexcept>
#include <algorithm>

namespace triangle {

flavor kind(double a, double b, double c) {
    const auto shortest = std::min({a, b, c});
    const auto longest = std::max({a, b, c});
    if (a + b + c < 2 * longest || shortest <= 0) {
        throw std::domain_error("Not a triangle!");
    }

    if (a == b && a == c) {
        return flavor::equilateral;
    }

    if (a == b || b == c || a == c) {
        return flavor::isosceles;
    }

    return flavor::scalene;
}

}  // namespace triangle
